File: phase-IV/frontend/src/lib/logging.ts
/**
 * Logging utility functions for the frontend application
 */

interface LogContext {
  timestamp: string;
  level: string;
  operation: string;
  context?: string;
  userId?: string;
  taskId?: string;
  errorMessage?: string;
}

/**
 * Log an operation with context
 */
export function log_operation(
  operation: string,
  context?: string,
  error_message?: string,
): void {
  const timestamp = new Date().toISOString();
  const logEntry: LogContext = {
    timestamp,
    level: error_message ? "ERROR" : "INFO",
    operation,
    ...(context && { context }),
    ...(error_message && { errorMessage: error_message }),
  };

  console.log(JSON.stringify(logEntry));
}

/**
 * Log an error with operation context
 */
export function log_error(error: Error, operation: string): void {
  const timestamp = new Date().toISOString();
  const logEntry: LogContext = {
    timestamp,
    level: "ERROR",
    operation: `ERROR_IN_${operation}`,
    errorMessage: error.message,
  };

  console.error(JSON.stringify(logEntry));
}

/**
 * Log authentication events
 */
export function log_authentication_event(
  event: string,
  user_id?: string,
  ip_address?: string,
): void {
  const timestamp = new Date().toISOString();
  const logEntry: LogContext = {
    timestamp,
    level: "INFO",
    operation: `AUTH_${event.toUpperCase()}`,
    ...(user_id && { userId: user_id }),
    ...(ip_address && { context: `IP: ${ip_address}` }),
  };

  console.log(JSON.stringify(logEntry));
}

/**
 * Log authorization decisions
 */
export function log_authorization_decision(
  action: string,
  user_id: string,
  resource: string,
  granted: boolean,
): void {
  const timestamp = new Date().toISOString();
  const decision = granted ? "GRANTED" : "DENIED";
  const logEntry: LogContext = {
    timestamp,
    level: granted ? "INFO" : "WARN",
    operation: `AUTHORIZATION_${decision.toUpperCase()}`,
    context: `${action} access to ${resource} for user ${user_id}`,
  };

  console.log(JSON.stringify(logEntry));
}

/**
 * Log token validation results
 */
export function log_token_validation_result(
  token_status: string,
  user_id?: string,
  reason?: string,
): void {
  const timestamp = new Date().toISOString();
  const logEntry: LogContext = {
    timestamp,
    level:
      token_status.includes("INVALID") || token_status.includes("EXPIRED")
        ? "WARN"
        : "INFO",
    operation: `TOKEN_${token_status.toUpperCase()}`,
    ...(user_id && { userId: user_id }),
    ...(reason && { context: `Reason: ${reason}` }),
  };

  console.log(JSON.stringify(logEntry));
}

/**
 * Log token lifecycle events
 */
export function log_token_lifecycle_event(
  event: string,
  user_id?: string,
  token_id?: string,
  details?: string,
): void {
  const timestamp = new Date().toISOString();
  const logEntry: LogContext = {
    timestamp,
    level: "INFO",
    operation: `TOKEN_LIFECYCLE_${event.toUpperCase()}`,
    ...(user_id && { userId: user_id }),
    ...(token_id && { context: `Token: ${token_id}` }),
    ...(details && { context: details }),
  };

  console.log(JSON.stringify(logEntry));
}

/**
 * Log security events
 */
export function log_security_event(
  event: string,
  user_id?: string,
  ip_address?: string,
  severity: string = "INFO",
): void {
  const timestamp = new Date().toISOString();
  const logEntry: LogContext = {
    timestamp,
    level: severity,
    operation: `SECURITY_${event.toUpperCase()}`,
    ...(user_id && { userId: user_id }),
    ...(ip_address && { context: `IP: ${ip_address}` }),
  };

  if (severity === "ERROR" || severity === "CRITICAL") {
    console.error(JSON.stringify(logEntry));
  } else {
    console.log(JSON.stringify(logEntry));
  }
}

File: phase-II/backend/src/core/logging.py
import logging
from datetime import datetime

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def log_operation(operation: str, user_id: str = None, task_id: int = None):
    """
    Log an operation with user and task context
    """
    timestamp = datetime.now().isoformat()
    context = f"[{timestamp}] Operation: {operation}"
    if user_id:
        context += f", User: {user_id}"
    if task_id:
        context += f", Task: {task_id}"

    logger.info(context)


def log_error(error: Exception, operation: str):
    """
    Log an error with operation context
    """
    timestamp = datetime.now().isoformat()
    logger.error(f"[{timestamp}] Error in '{operation}': {str(error)}")


def log_authentication_event(event: str, user_id: str = None, ip_address: str = None):
    """
    Log authentication-related events
    """
    timestamp = datetime.now().isoformat()
    context = f"[{timestamp}] Auth Event: {event}"
    if user_id:
        context += f", User: {user_id}"
    if ip_address:
        context += f", IP: {ip_address}"

    logger.info(context)


def log_authorization_decision(action: str, user_id: str, resource: str, granted: bool):
    """
    Log authorization decisions
    """
    timestamp = datetime.now().isoformat()
    decision = "GRANTED" if granted else "DENIED"
    context = f"[{timestamp}] Authorization {decision}: User {user_id} attempted to {action} {resource}"

    logger.info(context)


def log_token_validation_result(token_status: str, user_id: str = None, reason: str = None):
    """
    Log JWT token validation results
    """
    timestamp = datetime.now().isoformat()
    context = f"[{timestamp}] Token Validation: {token_status}"
    if user_id:
        context += f", User: {user_id}"
    if reason:
        context += f", Reason: {reason}"

    logger.info(context)


def log_token_lifecycle_event(event: str, user_id: str = None, token_id: str = None, details: str = None):
    """
    Log token lifecycle events (creation, refresh, expiry, etc.)
    """
    timestamp = datetime.now().isoformat()
    context = f"[{timestamp}] Token Lifecycle: {event}"
    if user_id:
        context += f", User: {user_id}"
    if token_id:
        context += f", Token: {token_id}"
    if details:
        context += f", Details: {details}"

    logger.info(context)


def log_security_event(event: str, user_id: str = None, ip_address: str = None, severity: str = "INFO"):
    """
    Log security-related events
    """
    timestamp = datetime.now().isoformat()
    context = f"[{timestamp}] Security Event [{severity}]: {event}"
    if user_id:
        context += f", User: {user_id}"
    if ip_address:
        context += f", IP: {ip_address}"

    if severity.upper() == "ERROR" or severity.upper() == "CRITICAL":
        logger.error(context)
    else:
        logger.info(context)